File: project_euler/problem_043/solution.py
def solve():
    """Find the sum of all 0-9 pandigital numbers with the sub-string
    divisibility property.

    Construction proceeds right-to-left, seeding with multiples of 17
    and extending leftward through divisors 13, 11, 7, 5, 3, 2.
    """
    primes = [17, 13, 11, 7, 5, 3, 2]

    # Seed: all 3-digit multiples of 17 with distinct digits
    candidates = []
    for m in range(17, 1000, 17):
        d8, d9, d10 = m // 100, (m // 10) % 10, m % 10
        if len({d8, d9, d10}) == 3:
            candidates.append(([d8, d9, d10], {d8, d9, d10}))

    # Extend leftward through each remaining prime divisor
    for p in primes[1:]:
        new_candidates = []
        for seq, used in candidates:
            for d in range(10):
                if d not in used:
                    if (100 * d + 10 * seq[0] + seq[1]) % p == 0:
                        new_candidates.append(([d] + seq, used | {d}))
        candidates = new_candidates

    # Assign remaining digit as d1 and compute sum
    total = 0
    for seq, used in candidates:
        d1 = (set(range(10)) - used).pop()
        total += int(''.join(map(str, [d1] + seq)))

    print(total)


solve()

File: project_euler/problem_043/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Right-to-left construction with constraint propagation.
    // Seed with multiples of 17, extend through 13, 11, 7, 5, 3, 2.
    int primes[] = {17, 13, 11, 7, 5, 3, 2};

    // Each candidate: (digit sequence, bitmask of used digits)
    vector<pair<vector<int>, int>> candidates;

    // Step 1: seed with 3-digit multiples of 17 with distinct digits
    for (int m = 17; m < 1000; m += 17) {
        int d8 = m / 100, d9 = (m / 10) % 10, d10 = m % 10;
        if (d8 != d9 && d8 != d10 && d9 != d10) {
            int mask = (1 << d8) | (1 << d9) | (1 << d10);
            candidates.push_back({{d8, d9, d10}, mask});
        }
    }

    // Steps 2-7: extend leftward
    for (int pi = 1; pi < 7; pi++) {
        int p = primes[pi];
        vector<pair<vector<int>, int>> next;
        for (auto& [seq, mask] : candidates) {
            for (int d = 0; d <= 9; d++) {
                if (mask & (1 << d)) continue;
                int val = 100 * d + 10 * seq[0] + seq[1];
                if (val % p == 0) {
                    vector<int> new_seq = {d};
                    new_seq.insert(new_seq.end(), seq.begin(), seq.end());
                    next.push_back({new_seq, mask | (1 << d)});
                }
            }
        }
        candidates = next;
    }

    // Step 8: assign remaining digit as d1
    long long total = 0;
    for (auto& [seq, mask] : candidates) {
        for (int d = 0; d <= 9; d++) {
            if (!(mask & (1 << d))) {
                long long num = d;
                for (int x : seq) num = num * 10 + x;
                total += num;
            }
        }
    }

    cout << total << endl;
    return 0;
}
